resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/rest/kubernetes/StagedResourcesCleaner.scala

```diff
@@ -0,0 +1,154 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.spark.deploy.rest.kubernetes
+
+import java.util.concurrent.{ScheduledExecutorService, TimeUnit}
+
+import scala.collection.JavaConverters._
+import scala.collection.mutable
+
+import org.apache.spark.SparkException
+import org.apache.spark.internal.Logging
+import org.apache.spark.util.{Clock, Utils}
+
+private[spark] trait StagedResourcesCleaner {
+
+  def start(): Unit
+
+  def registerResourceForCleaning(
+      resourceId: String, stagedResourceOwner: StagedResourcesOwner): Unit
+
+  def markResourceAsUsed(resourceId: String): Unit
+}
+
+private class StagedResourcesCleanerImpl(
+      stagedResourcesStore: StagedResourcesStore,
+      kubernetesClientProvider: ResourceStagingServiceKubernetesClientProvider,
+      cleanupExecutorService: ScheduledExecutorService,
+      clock: Clock,
+      initialAccessExpirationMs: Long,
+      resourceCleanupIntervalMs: Long)
+    extends StagedResourcesCleaner {
+
+  private val RESOURCE_LOCK = new Object()
+  private val activeResources = mutable.Map.empty[String, MonitoredResource]
+  private val unusedResources = mutable.Map.empty[String, UnusedMonitoredResource]
+
+  override def start(): Unit = {
+    cleanupExecutorService.scheduleAtFixedRate(
+        new CleanupRunnable(),
+        resourceCleanupIntervalMs,
+        resourceCleanupIntervalMs,
+        TimeUnit.MILLISECONDS)
+  }
+
+  override def registerResourceForCleaning(
+      resourceId: String, stagedResourceOwner: StagedResourcesOwner): Unit = {
+    RESOURCE_LOCK.synchronized {
+      unusedResources(resourceId) = UnusedMonitoredResource(
+          clock.getTimeMillis() + initialAccessExpirationMs,
+          MonitoredResource(resourceId, stagedResourceOwner))
+
+    }
+  }
+
+  override def markResourceAsUsed(resourceId: String): Unit = RESOURCE_LOCK.synchronized {
+    val resource = unusedResources.remove(resourceId)
+    resource.foreach { res =>
+      activeResources(resourceId) = res.resource
+    }
+  }
+
+  private class CleanupRunnable extends Runnable with Logging {
+
+    override def run(): Unit = {
+      // Make a copy so we can iterate through this while modifying
+      val activeResourcesCopy = RESOURCE_LOCK.synchronized {
+        Map.apply(activeResources.toSeq: _*)
+      }
+      for ((resourceId, resource) <- activeResourcesCopy) {
+        Utils.tryWithResource(
+            kubernetesClientProvider.getKubernetesClient(
+                resource.stagedResourceOwner.ownerNamespace,
+                resource.stagedResourceOwner.ownerMonitoringCredentials)) { kubernetesClient =>
+          val namespace = kubernetesClient.namespaces()
+              .withName(resource.stagedResourceOwner.ownerNamespace)
+              .get()
+          if (namespace == null) {
+            logInfo(s"Resource with id $resourceId is being removed. The owner's namespace" +
```

Comment:
`is being removed` -> `No longer tracking resource with id $resourceId`

At first read I thought it was being deleted out of k8s, rather than forgotten in RSS